libs/base/Decidable/Order.idr

```diff
@@ -0,0 +1,156 @@
+module Decidable.Order
+
+import Decidable.Decidable
+import Decidable.Equality
+import Data.Fin
+import Data.Fun
+import Data.Nat
+import Data.Rel
+
+%hide Prelude.Equal
+
+--------------------------------------------------------------------------------
+-- Utility Lemmas
+--------------------------------------------------------------------------------
+
+--------------------------------------------------------------------------------
+-- Preorders, Posets, total Orders, Equivalencies, Congruencies
+--------------------------------------------------------------------------------
+
+interface Preorder t (po : t -> t -> Type) where
+  total transitive : (a : t) -> (b : t) -> (c : t) -> po a b -> po b c -> po a c
+  total reflexive : (a : t) -> po a a
+
+interface (Preorder t po) => Poset t (po : t -> t -> Type) where
+  total antisymmetric : (a : t) -> (b : t) -> po a b -> po b a -> a = b
+
+interface (Poset t to) => Ordered t (to : t -> t -> Type) where
+  total order : (a : t) -> (b : t) -> Either (to a b) (to b a)
+
+interface (Preorder t eq) => Equivalence t (eq : t -> t -> Type) where
+  total symmetric : (a : t) -> (b : t) -> eq a b -> eq b a
+
+interface (Equivalence t eq) => Congruence t (f : t -> t) (eq : t -> t -> Type) where
+  total congruent : (a : t) ->
+                    (b : t) ->
+                    eq a b ->
+                    eq (f a) (f b)
+
+minimum : (Ordered t to) => t -> t -> t
+minimum {to} x y with (order {to} x y)
```

Comment:
You seem to consistently need to introduce `to` and apply it
explicitly which makes sense given that their can be multiple
interesting orders on a given type.

I think this is a sign `to` should be an explicit argument in
all of these definitions.